src/mqtt/protocol/packets/packets.cr

```diff
@@ -0,0 +1,85 @@
+require "../io"
+
+macro decode_assert(condition, err, *args)
+  {% if (err.class_name == "StringLiteral" || err.class_name == "StringInterpolation") %}
+    # err is a string
+    ({{condition}} || raise Error::PacketDecode.new {{err}})
+  {% elsif (err.class_name == "Call") %}
+    # err is a call that we assume returns a string e.g. sprintf()
+    ({{condition}} || raise Error::PacketDecode.new {{err}})
+  {% else %}
+    # here we just assume it's a class name
+    ({{condition}} || raise {{err}}.new({{*args}}))
+  {% end %}
+end
+
+module MQTT
+  module Protocol
+    alias Flags = UInt8
+
+    abstract struct Packet
+      abstract def to_io(io : MQTT::Protocol::IO)
+
+      @remaining_length : UInt32 = 2
+
+      def remaining_length : UInt32
+        @remaining_length
+      end
+
+      def bytesize : UInt32
+        remaining_length + 2
```

Comment:
I don't think this is this easy? Shouldn't it be like `remaining_lenght + bytesize_of_remaining_lenght + 1` where 1 is the fixed header size?
```suggestion
        if remaining_length < 127
          remaining_length + 1 + 1
        elsif remaining_length < 16_383
          remaining_length + 2 + 1 
        elsif remaining_length < 2_097_151
          remaining_length + 3 + 1          
        else
          remaining_length + 4 + 1          
        end
```
Something like this?